### PR TITLE
ENH: Serializable acal

### PR DESCRIPTION
**Pull request by NoahRJohnson**
- Moved over some acal bindings from Terra DSM's geo_corr module
- Added pickling support for acal_match_graph

acal has several objects that reference each other, via shared_ptr and raw pointers. This is hard to serialize, so instead I assign an ID to each object, and replace references to the objects with their IDs. Then when deserializing, I rebuild the objects and replace the IDs by pointers to these new objects. This adds some amount of overhead when constructing match graphs in parallel.

"If it’s stupid but it works, it isn’t stupid."

### CODE REVIEW

**Comment by NoahRJohnson** [comment]
Waiting on VXL PR #779

**Comment by decrispell** [comment]
The VXL PR has now been merged.

**Comment by drewgilliam** [comment]
- much of pystruct is already available - see pyvxl_utils at the top level

- can you explain why the serialization routine should be here and not in acal proper?  

- can you please add some unit tests?

**Comment by NoahRJohnson** [comment]
- Thanks, I've removed py_struct in favor of using pyvxl_utils

- The proper VXL way to serialize objects is with VSL io, right? I figure that would have a totally different implementation, as would any other serialization framework we might want to use in the future. Those future implementations would ideally keep track of unique pointer IDs internally as they serialize, instead of copying data into new representation structures. The representation tuple built here for pickle adds a lot of overhead, and makes assumptions about what structures pybind11 can automatically handle.

- Tests added, except for the acal_match_tree binding. I found a bug there: instantiating and then deleting a match_tree object causes a segfault. I've no clue what's causing that - any ideas?

**Comment by drewgilliam** [comment]
IMO, serialization belongs in VXL - having a method to serialize `acal_match_graph` would be useful for more than just pyvxl pickling.  I've added `acal_io` via vxl/vxl#780 that includes vsl serialization and equality operators for `acal_corr`, `acal_match_pair`, and `match_params`.  These should be exploited here for pickling, `py::self == py::self`, and unit tests.

I left placeholders in VXL for `acal_match_graph` serialization & unit tests.  The code here is likely useful in VXL, swapping the tuple-based write/read to use vsl utilities.  It's also advisable to have an equality operator in VXL to permit easy unit tests on the write/read operation (rather than the large unit test for equality you have in pyvxl).

I'll look into your `acal_match_tree` bug.

File: contrib/acal/pyacal.cxx [hunk 1]
```diff
@@ -1,24 +1,34 @@
...
 #include <pybind11/stl.h>
-#include <pybind11/numpy.h>
 
 #include <bpgl/acal/acal_f_utils.h>
 #include <bpgl/acal/acal_match_graph.h>
+#include <vpgl/vpgl_affine_camera.h>
 
 namespace py = pybind11;
 
+using namespace py::literals;
```
Comment (drewgilliam) [inline contrib/acal/pyacal.cxx:19]: Why do you need py::literals?

File: test/test_pyacal.py [hunk 2]
```diff
@@ -44,12 +50,219 @@ def __init__(self, *args, **kwargs):
     }
 
 
+class acal_corr(AcalBase, unittest.TestCase):
```
Comment (drewgilliam) [inline test/test_pyacal.py:53]: I'd like to see all pickle operations validated

**Comment by drewgilliam** [comment]
After discussion with @NoahRJohnson, closing this PR for now.  

PR #91 links to a few new acal_io bindings, `acal_match_graph` serialization will be moved to acal_io and new pyvxl bindings will be made at that time.

**Pull request closed without merging**